presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,255 @@
+                    sourceStats.stream().map(VariableStatsEstimate::getHighValue).collect(Collectors.toList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.maxValue())));
+            minValue = firstNonNaN(minValue, processSingleArgumentStatistic(
+                    sourceStats.stream().map(VariableStatsEstimate::getLowValue).collect(Collectors.toList()),
+                    paramIndexVsStatsMap.getKey(),
+                    applyPropagateAllStats(propagateAllStats, scalarPropagateSourceStats.minValue())));
+        }
+        if (!isFinite(maxValue) || !isFinite(minValue)) {
+            minValue = NEGATIVE_INFINITY;
+            maxValue = POSITIVE_INFINITY;
+        }
+        return VariableStatsEstimate.builder()
+                .setLowValue(minValue)
+                .setHighValue(maxValue)
+                .setNullsFraction(nullFraction)
+                .setAverageRowSize(minExcludingNaNs(averageRowSize, getReturnTypeWidth(callExpression, UNKNOWN)))
+                .setDistinctValuesCount(distinctValuesCount).build();
+    }
+
+    private double processDistinctValuesCount(List<Double> sourceStats, int sourceStatsArgumentIndex, StatsPropagationBehavior operation)
+    {
+        double distinctValuesCount =
+                processSingleArgumentStatistic(sourceStats, sourceStatsArgumentIndex, operation);
+        if (isFinite(outputRowCount)) {
+            distinctValuesCount = min(distinctValuesCount, outputRowCount * (1 - firstNonNaN(nullFraction, 0.0)));
+        }
+        return distinctValuesCount;
+    }
+
+    private double processAvgRowSize(List<Double> sourceStats, int sourceStatsArgumentIndex, StatsPropagationBehavior operation)
+    {
+        double avgRowSize =
+                processSingleArgumentStatistic(sourceStats, sourceStatsArgumentIndex, operation);
+        return minExcludingNaNs(avgRowSize, getReturnTypeWidth(callExpression, operation));
+    }
+
+    private double processSingleArgumentStatistic(
+            List<Double> sourceStats,
+            int sourceStatsArgumentIndex,
+            StatsPropagationBehavior operation)
+    {
+        // sourceStatsArgumentIndex is index of the argument on which
+        // ScalarPropagateSourceStats annotation was applied.
+        double statValue = NaN;
+        if (!operation.isSingleArgumentStats()) {
+            boolean sourceStatUnknown = false;
+            for (int i = 0; i < sourceStats.size() && !sourceStatUnknown; i++) {
+                switch (operation) {
+                    case MAX_TYPE_WIDTH_VARCHAR:
+                        statValue = maxExcludingNaNs(statValue, typeWidthVarchar(callExpression, i));
+                        break;
+                    case USE_MIN_ARGUMENT:
+                        sourceStatUnknown = !isFinite(sourceStats.get(i));
+                        statValue = minExcludingNaNs(statValue, sourceStats.get(i));
```

Comment:
Instead of tracking `sourceStatUnknown` we could just return `NaN` from here if `sourceStats.get(i)` is a `NaN`

Reply:
I considered that, 

* This way it is explicit, the reader of the code will know our intention.
* If I use min instead of minExcludingNaNs, then I need to add condition if statValue is NaN and sourceStats are defined then based on whether it is USE_MIN_ARGUMENT, USE_MAX... assign the initial value of statValue.
For the sake of conciseness and clarity, I leaned this way.  However, It may be otherwise, I am happy to change it - WDYT?